src/sources/generator.rs

```diff
@@ -3,34 +3,129 @@ use crate::{
     event::Event,
     internal_events::GeneratorEventProcessed,
     shutdown::ShutdownSignal,
+    sources::util::fake::{apache_common_log_line, apache_error_log_line, syslog_5424_log_line},
     Pipeline,
 };
 use futures::{stream::StreamExt, SinkExt};
+use rand::seq::SliceRandom;
 use serde::{Deserialize, Serialize};
+use snafu::Snafu;
 use std::task::Poll;
 use tokio::time::{interval, Duration};
 
 #[derive(Clone, Debug, Default, Deserialize, Serialize)]
-#[serde(deny_unknown_fields)]
 pub struct GeneratorConfig {
-    #[serde(default)]
-    sequence: bool,
-    lines: Vec<String>,
-    #[serde(default)]
-    batch_interval: Option<f64>,
+    #[serde(alias = "batch_interval")]
+    interval: Option<f64>,
     #[serde(default = "usize::max_value")]
     count: usize,
+    #[serde(flatten)]
+    format: OutputFormat,
+}
+
+#[derive(Debug, PartialEq, Snafu)]
+pub enum GeneratorConfigError {
+    #[snafu(display("A non-empty list of lines is required for the random format"))]
+    RandomGeneratorItemsEmpty,
+}
+
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(tag = "format", rename_all = "snake_case")]
+pub enum OutputFormat {
+    #[derivative(Default)]
+    Random {
+        #[serde(default)]
+        sequence: bool,
+        lines: Vec<String>,
+    },
+    ApacheCommon,
+    ApacheError,
+    #[serde(alias = "rfc5424")]
+    Syslog,
+}
+
+impl OutputFormat {
+    fn generate_event(&self, n: usize) -> Event {
+        emit!(GeneratorEventProcessed);
+
+        let line = match self {
+            Self::Random {
+                sequence,
+                ref lines,
+            } => Self::random_line_generate(*sequence, lines, n),
+            Self::ApacheCommon => apache_common_log_line(),
+            Self::ApacheError => apache_error_log_line(),
+            Self::Syslog => syslog_5424_log_line(),
+        };
+        Event::from(line)
+    }
+
+    fn random_line_generate(sequence: bool, lines: &[String], n: usize) -> String {
+        // unwrap can be called here because lines cannot be empty
+        let line = lines.choose(&mut rand::thread_rng()).unwrap().into();
+
+        if sequence {
+            format!("{} {}", n, line)
+        } else {
+            line
+        }
```

Comment:
I would still recommend moving the `line.into()` into the conditional, otherwise lines with sequence numbers cause an extraneous allocation.

Reply:
Ah, I missed that. Fixed.